File: app/src/main/java/mohamed/atef/mondiatask/repositories/SearchRepository.java
package mohamed.atef.mondiatask.repositories;

import android.content.Context;
import android.util.Log;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.net.HttpURLConnection;
import java.net.URL;

import mohamed.atef.mondiatask.AppUtils;
import mohamed.atef.mondiatask.R;
import mohamed.atef.mondiatask.models.ClientTokenModel;

public class SearchRepository {
    InputStream inputStream = null;
    private BufferedReader reader;

    public String makeSearchCall(String jsonBody, Context mContext, ClientTokenModel clientTokenModel) {
        String UsersDesires_JsonSTR = null;
        HttpURLConnection urlConnection = null;
        try {
            URL url = new URL(AppUtils.Companion.getSearchEnd());
            urlConnection = (HttpURLConnection) url.openConnection();
            urlConnection.setDoInput(true);
            urlConnection.setDoOutput(true);
            urlConnection.setRequestMethod("GET");
            urlConnection.addRequestProperty("Authorization", clientTokenModel.getTokenType() + " " + clientTokenModel.getAccessToken());
            urlConnection.addRequestProperty("Content-Type", "application/json");
            urlConnection.addRequestProperty("Accept", "application/json");
            urlConnection.addRequestProperty("X-MM-GATEWAY-KEY", mContext.getString(R.string.gateway_key));

            try {
                OutputStreamWriter streamWriter = new OutputStreamWriter(urlConnection.getOutputStream());
                streamWriter.write(jsonBody);
                streamWriter.flush();
            } catch (Exception e) {
                e.printStackTrace();
            }

            if (urlConnection.getResponseCode() != HttpURLConnection.HTTP_OK) {
                Log.e("test", urlConnection.getResponseMessage());
                return null;
            } else {
                StringBuffer buffer = new StringBuffer();
                if (inputStream == null) {
                    UsersDesires_JsonSTR = null;
                }
                try{
                    reader = new BufferedReader(new InputStreamReader(urlConnection.getInputStream()));
                    String line;
                    if (reader != null) {
                        while ((line = reader.readLine()) != null) {
                            buffer.append(line + "\n");
                        }
                    }
                }catch (Exception e){
                    e.printStackTrace();
                }
                if (buffer.length() == 0) {
                    return null;
                }
                UsersDesires_JsonSTR = buffer.toString();
                Log.v("error:", "Articles String: " + UsersDesires_JsonSTR);
                return null;
            }
        } catch (IOException e) {
            Log.e("error:", "Error here Exactly ", e);
            return null;
        } finally {
            if (urlConnection != null) {
                urlConnection.disconnect();
            }
            if (reader != null) {
                try {
                    reader.close();
                } catch (final IOException e) {
                    Log.e("error:", "Error closing stream", e);
                }
            }
        }
    }


    public String returnClientToken(Context mContext) {
        String UsersDesires_JsonSTR = null;
        HttpURLConnection urlConnection = null;
        try {
            URL url = new URL(AppUtils.Companion.getClientTokenEnd());
            urlConnection = (HttpURLConnection) url.openConnection();
            urlConnection.setDoInput(true);
            urlConnection.setDoOutput(true);
            urlConnection.setRequestMethod("POST");
            urlConnection.addRequestProperty("Content-Type", "application/x-www-form-urlencoded");
            urlConnection.addRequestProperty("Accept", "application/json");
            urlConnection.addRequestProperty("X-MM-GATEWAY-KEY", mContext.getString(R.string.gateway_key));
            urlConnection.connect();

            if (urlConnection.getResponseCode() != HttpURLConnection.HTTP_OK) {
                Log.e("test", urlConnection.getResponseMessage());
                return null;
            } else {
                StringBuffer buffer = new StringBuffer();
                reader = new BufferedReader(new InputStreamReader(urlConnection.getInputStream()));
                String line;
                if (reader != null) {
                    while ((line = reader.readLine()) != null) {
                        buffer.append(line + "\n");
                    }
                }
                if (buffer.length() == 0) {
                    return null;
                }
                UsersDesires_JsonSTR = buffer.toString();
                Log.v("error:", "Articles String: " + UsersDesires_JsonSTR);
                return UsersDesires_JsonSTR;
            }
        } catch (IOException e) {
            Log.e("error:", "Error here Exactly ", e);
            return null;
        } finally {
            if (urlConnection != null) {
                urlConnection.disconnect();
            }
            if (reader != null) {
                try {
                    reader.close();
                } catch (final IOException e) {
                    Log.e("error:", "Error closing stream", e);
                }
            }
        }
    }
}

File: app/src/main/java/mohamed/atef/mondiatask/repositories/SearchRepository.kt
package mohamed.atef.mondiatask.repositories

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import mohamed.atef.mondiatask.AppUtils.Companion.clientToken
import mohamed.atef.mondiatask.AppUtils.Companion.searchEnd
import mohamed.atef.mondiatask.models.SearchResults
import mohamed.atef.mondiatask.parser.SearchParser
import org.json.JSONException
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL

class SearchRepository (private val responseParser: SearchParser){

    private lateinit var UsersDesires_JsonSTR: String
    private lateinit var reader: BufferedReader

    fun makeSearchCall(
        jsonBody:String
    ):String?{
        val url = URL(searchEnd)
        (url.openConnection() as? HttpURLConnection)?.run {
            requestMethod = "GET"
            setRequestProperty("Content-Type", "application/json; utf-8")
            setRequestProperty("Accept", "application/json")
            setRequestProperty("X-MM-GATEWAY-KEY", "G2269608a-bf41-2dc7-cfea-856957fcab1e")
            setRequestProperty("Authorization", "Bearer"+ clientToken)
            doOutput = true
            outputStream.write(jsonBody.toByteArray())
            try {
                val buffer = StringBuffer()
                reader = BufferedReader(InputStreamReader(inputStream))
                var line: String
                if (reader != null) {
                    while (reader.readLine().also { line = it } != null) {
                        buffer.append(
                            """
                $line
                """.trimIndent()
                        )
                    }
                }
                if (buffer.length == 0) {
                    Log.v("Error:", "error ... ")
                    return null
                }
                UsersDesires_JsonSTR = buffer.toString();
                Log.v("JSON:", "Articles String: " + UsersDesires_JsonSTR);
            }catch (e: IOException){
                Log.e("Error:", "Error here Exactly ", e);
                return null
            } finally {
                if (url != null) {
                    (url as? HttpURLConnection)?.disconnect()
                }
                if (reader != null) {
                    try {
                        reader.close();
                    } catch (e: IOException) {
                        Log.e("Error:", "Error closing stream", e);
                        return null
                    }
                }
            }
            try {
                return UsersDesires_JsonSTR
            }catch (e: JSONException){
                Log.e("Error:", "didn't got Users Desires from getJsonData method", e);
                e.printStackTrace();
                return null
            }
        }
        return UsersDesires_JsonSTR
    }
}
